atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleAlterTableIntegrationTest.java

```diff
@@ -0,0 +1,176 @@
+    private static final TableReference TABLE_REFERENCE = TableReference.create(NAMESPACE, "foo");
+    private static final DbKeyValueServiceConfig CONFIG_WITH_ALTER = ImmutableDbKeyValueServiceConfig.builder()
+            .from(DbKvsOracleTestSuite.getKvsConfig())
+            .ddl(ImmutableOracleDdlConfig.builder()
+                    .overflowMigrationState(OverflowMigrationState.FINISHED)
+                    .addAlterTablesOrMetadataToMatch(TABLE_REFERENCE)
+                    .build())
+            .build();
+
+    private static final String COLUMN_NAME = "variable";
+
+    private static final TableMetadata EXPECTED_TABLE_METADATA = TableMetadata.builder()
+            .singleRowComponent("name", ValueType.STRING)
+            .singleNamedColumn("var", COLUMN_NAME, ValueType.STRING)
+            .build();
+
+    private static final TableMetadata OLD_TABLE_METADATA = TableMetadata.builder()
+            .singleRowComponent("name", ValueType.STRING)
+            .singleNamedColumn("var", COLUMN_NAME, ValueType.FIXED_LONG)
+            .build();
+
+    private static final Cell DEFAULT_CELL_1 = createCell("foo", "bar");
+    private static final Cell DEFAULT_CELL_2 = createCell("something", "bar");
+
+    private static final byte[] DEFAULT_VALUE = PtBytes.toBytes("amazing");
+    private static final long TIMESTAMP_1 = 1L;
+    private static final long TIMESTAMP_2 = 2L;
+
+    private static final Map<Cell, byte[]> ROW_1 = Map.of(DEFAULT_CELL_1, DEFAULT_VALUE);
+    private static final Map<Cell, byte[]> ROW_2 = Map.of(DEFAULT_CELL_2, DEFAULT_VALUE);
+
+    private KeyValueService defaultKvs;
+    private ConnectionSupplier connectionSupplier;
+    private OracleTableNameGetter oracleTableNameGetter;
+
+    @Before
+    public void before() {
+        defaultKvs = TRM.getDefaultKvs();
+        connectionSupplier = DbKvsOracleTestSuite.getConnectionSupplier(defaultKvs);
+        oracleTableNameGetter = OracleTableNameGetterImpl.createDefault(
+                (OracleDdlConfig) DbKvsOracleTestSuite.getKvsConfig().ddl());
+    }
+
+    @After
+    public void after() {
+        defaultKvs.dropTables(defaultKvs.getAllTableNames());
+        defaultKvs.dropTable(DbKvsOracleTestSuite.getKvsConfig().ddl().metadataTable());
+        connectionSupplier.close();
+    }
+
+    @Test
+    public void whenConfiguredAlterTableToMatchMetadataAndOldDataIsStillReadable() {
+        defaultKvs.createTable(TABLE_REFERENCE, EXPECTED_TABLE_METADATA.persistToBytes());
+        writeData(defaultKvs, ROW_1, TIMESTAMP_1);
+        assertThatDataCanBeRead(defaultKvs, DEFAULT_CELL_1, TIMESTAMP_1);
+
+        dropOverflowColumn();
+        defaultKvs.putMetadataForTable(TABLE_REFERENCE, OLD_TABLE_METADATA.persistToBytes());
+        assertThatThrownBy(() -> fetchData(defaultKvs, DEFAULT_CELL_1, TIMESTAMP_1));
+
+        KeyValueService workingKvs = ConnectionManagerAwareDbKvs.create(CONFIG_WITH_ALTER);
```

Comment:
Don't forget to close your KVS :) You can add it to TRM.registerKvs which will also handle closing it (and then you can have a utility method like create KVS that registers then returns the KVS)